src/library/equations_compiler/elim_match.cpp

```diff
@@ -502,17 +503,14 @@ struct elim_match_fn {
                 expr const & p = head(eqn.m_patterns);
                 if (is_local(p)) {
                     has_variable = true; return true;
-                } else if (is_constructor_app(p)) {
+                }
+                type_context ctx = mk_type_context(eqn.m_lctx);
+                bool is_val = is_value(ctx, p);
+                if ((is_constructor_app(p) && (value_is_contructor || !is_val)) ||
+                    (value_is_contructor && is_val)) {
                     has_constructor = true; return true;
-                } else {
-                    type_context ctx = mk_type_context(eqn.m_lctx);
-                    if (is_value(ctx, p)) {
-                        if (value_is_contructor) has_constructor = true;
-                        return true;
-                    } else {
-                        return false;
-                    }
                 }
+                return false;
             });
         return r && has_variable && has_constructor;
     }
```

Comment:
> When we use "value" transition, the equation lemmas do not hold definitionally.

BTW, this is why the test `eq11.lean` failed here:
https://ci.appveyor.com/project/leodemoura/lean/build/1.0.634